### ISSUE TITLE
Bug in calc_pmk() function causes wrong PMK to be computed

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#215]
**Reported by ebfe on 24 May 2007 21:52 UTC**

a bug in calc_pmk() causes the pmk to be wrong if the essid parameter does not point to a 33+4 character string. the function assumes the bytes behind the string to be 0. a function should never try to read behind a string's zero-terminator.

the patch fixes airolib-ng's ability to calculate correct PMKs

```
--- src/crypto.c	(revision 437)
+++ src/crypto.c	(working copy)
@@ -152,18 +152,22 @@
 
 /* derive the PMK from the passphrase and the essid */
 
-void calc_pmk( char *key, char *essid, uchar pmk[)
+void calc_pmk( char *key, char *essid_pre, uchar pmk[40](40]) )
 {
 	int i, j, slen;
 	uchar buffer[essid[33+4](65];
+	uchar);
 	SHA_CTX ctx_ipad;
 	SHA_CTX ctx_opad;
 	SHA_CTX sha1_ctx;
 
-	slen = strlen( essid ) + 4;
+	memset(essid,0,sizeof(essid));
+	memcpy(essid,essid_pre,strlen(essid_pre));
+	slen = strlen(essid)+4;
 
 	/* setup the inner and outer contexts */
-
 	memset( buffer, 0, sizeof( buffer ) );
 	strncpy( (char *) buffer, key, sizeof( buffer ) - 1 );

```

**Comment by aircrack-ng** [comment]
**Comment by misterx on 24 May 2007 21:52 UTC**

cosmetic changes to the description

**Comment by aircrack-ng** [comment]
**Comment by misterx on 24 May 2007 21:52 UTC**

(In #477) Fixed bug in calc_pmk() function causes wrong PMK to be computed (Closes: #215)

**Comment by aircrack-ng** [comment]
**Comment by misterx on 24 May 2007 21:52 UTC**

Apply that to airdecap-ng too.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 24 May 2007 21:52 UTC**

(In #535) airdecap-ng: Fixed bug in calc_pmk() function that causes wrong PMK to be computed (Closes: #215)
